File: lib/dynamodb_construct.py
"""DynamoDB table for payment transactions."""
from constructs import Construct
from cdktf_cdktf_provider_aws.dynamodb_table import DynamodbTable


class DynamoDbConstruct(Construct):
    """DynamoDB table with GSI and stream."""

    def __init__(self, scope: Construct, id: str, environment_suffix: str, kms_key_id: str):
        super().__init__(scope, id)

        self.table = DynamodbTable(
            self, "payment-transactions",
            name=f"payment-transactions-{environment_suffix}",
            billing_mode="PAY_PER_REQUEST",
            hash_key="transaction_id",
            attribute=[
                {"name": "transaction_id", "type": "S"},
                {"name": "customer_id", "type": "S"}
            ],
            global_secondary_index=[
                {
                    "name": "customer-index",
                    "hashKey": "customer_id",
                    "projectionType": "ALL"
                }
            ],
            stream_enabled=True,
            stream_view_type="NEW_AND_OLD_IMAGES",
            point_in_time_recovery={"enabled": True},
            server_side_encryption={
                "enabled": True,
                "kms_key_arn": kms_key_id
            },
            tags={
                "Name": f"payment-transactions-{environment_suffix}",
                "Environment": environment_suffix,
                "Project": "payment-processing",
                "CostCenter": "engineering"
            }
        )

        self.stream_arn = self.table.stream_arn

File: lib/dynamodb-construct.ts
import { Construct } from 'constructs';
import { DynamodbTable } from '@cdktf/provider-aws/lib/dynamodb-table';
import { SecurityConstruct } from './security-construct';

interface DynamoDbConstructProps {
  prefix: string;
  security: SecurityConstruct;
}

export class DynamoDbConstruct extends Construct {
  constructor(scope: Construct, id: string, props: DynamoDbConstructProps) {
    super(scope, id);
    // For each region, create DynamoDB table encrypted with KMS
    Object.keys(props.security.kmsKeys).forEach(region => {
      const kmsKey = props.security.kmsKeys[region];
      const tableSuffix = Math.random().toString(36).substring(2, 8);
      // Import DynamoDBTable from @cdktf/provider-aws
      // Dynamically require to avoid import errors if not used elsewhere
      new DynamodbTable(
        this,
        `${props.prefix}-dynamodb-table-${region}-${tableSuffix}`,
        {
          provider: kmsKey.provider,
          name: `${props.prefix}-table-${region}-${tableSuffix}`,
          billingMode: 'PAY_PER_REQUEST',
          hashKey: 'id',
          attribute: [{ name: 'id', type: 'S' }],
          serverSideEncryption: {
            enabled: true,
            kmsKeyArn: kmsKey.arn,
          },
          tags: {
            Name: `${props.prefix}-dynamodb-table-${region}-${tableSuffix}`,
            Environment: props.prefix,
          },
        }
      );
    });
  }
}
